src/uct/sm/mm/base/mm_ep.c

```diff
@@ -281,7 +284,9 @@ static UCS_F_ALWAYS_INLINE ssize_t uct_mm_ep_am_common_send(
                                                     &ep->arb_group);
                 UCS_STATS_UPDATE_COUNTER(ep->super.stats, UCT_EP_STAT_NO_RES,
                                          1);
-                return UCS_ERR_NO_RESOURCE;
+                status = uct_mm_ep_peer_check(ep, flags);
+                return ucs_likely(status == UCS_OK) ? UCS_ERR_NO_RESOURCE :
+                       status;
```

Comment:
1. what about line 275?
2. i'm not sure compiler will be able to optimize away all these branches. maybe we can simplify somehow?

Reply:
> what about line 275?

I thought that it is not needed, but I agree better to do PEER_CHECK here too
fixed

> i'm not sure compiler will be able to optimize away all these branches. maybe we can simplify somehow?

agree, fixed now - can just do `ucs_assertv_always` inside the handling of `PEER_CHECK` flag.

Reply:
ucs_assertv_always is also a branch..
WDYT about changing "uct_iface_schedule_ep_err" so it will not allocate err_info, but use "ep" as context, so it can return void? then uct_ep_keepalive_check() can also return void, and we can call it directly.
We can assume the status in uct_iface_ep_error_handle_progress will always be known (actually, better change it from UCS_ERR_ENDPOINT_TIMEOUT to CONN_RESET)

Reply:
> ucs_assertv_always is also a branch..

but this branch only when doing PEER_CHECK

> WDYT about changing "uct_iface_schedule_ep_err" so it will not allocate err_info, but use "ep" as context, so it can return void? then uct_ep_keepalive_check() can also return void, and we can call it directly.

yes, it sounds as good idea, done

> We can assume the status in uct_iface_ep_error_handle_progress will always be known (actually, better change it from UCS_ERR_ENDPOINT_TIMEOUT to CONN_RESET)

done